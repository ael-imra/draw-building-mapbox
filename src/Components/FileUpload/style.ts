export const sx = {
  uploadButton: {
    width: '85%',
    height: '45px',
    backgroundColor: '#60d2db',
    borderRadius: '12px',
    color: 'white',
    fontSize: '17px',
    fontWeight: 600,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    textTransform: 'uppercase',
    cursor: 'pointer',
  },
  error: {
    fontSize: 20,
    color: '#D1111B',
    fontWeight: 500,
    marginBottom: '10px',
  },
};
