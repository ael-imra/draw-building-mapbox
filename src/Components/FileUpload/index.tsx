import { Grid, TextField } from '@mui/material';
import { useState } from 'react';
import { sx } from './style';

interface FileUploadProps {
  title: string;
  onDataChange: Function;
}

const FileUpload = (props: FileUploadProps) => {
  const { title, onDataChange } = props;
  const [error, setError] = useState('');

  const handleFileChange = (event: any) => {
    const file = event?.target?.files?.[0];
    if (file) {
      const reader = new FileReader();
      reader.onload = (data: any) => {
        try {
          if (data?.target?.result) onDataChange(JSON.parse(data.target.result));
        } catch (error) {
          setError("Can't load this file");
        }
      };
      reader.readAsText(file);
    }
  };
  const handleChooseFile = () => {
    const inputFile = document.getElementById('file');
    if (inputFile) {
      inputFile.click();
    }
  };

  return (
    <>
      {error && <Grid sx={sx.error}>{error}</Grid>}
      <TextField id="file" type="file" sx={{ display: 'none' }} onChange={handleFileChange} />
      <Grid sx={sx.uploadButton} onClick={handleChooseFile}>
        {title}
      </Grid>
    </>
  );
};

export default FileUpload;
