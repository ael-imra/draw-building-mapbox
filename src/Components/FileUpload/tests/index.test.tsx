import { render, fireEvent, screen, waitFor, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import FileUpload from '../';

afterEach(cleanup);
describe('FileUpload', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<FileUpload title={'title'} onDataChange={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the title', () => {
    const title = 'Choose File';

    render(<FileUpload title={title} onDataChange={() => {}} />);

    expect(screen.getByText(title)).toBeInTheDocument();
  });

  it('should call onDataChange when a file is uploaded', async () => {
    const onDataChange = jest.fn();

    render(<FileUpload title="Choose File" onDataChange={onDataChange} />);
    const file = screen.getByTestId('file');
    fireEvent.change(file, { target: { files: [new File(['{}'], 'test.geojson')] } });

    await waitFor(() => {
      expect(onDataChange).toHaveBeenCalled();
    });
  });

  it('should show an error message when an invalid file is uploaded', async () => {
    const errorMessage = "Can't load this file";

    render(<FileUpload title="Choose File" onDataChange={() => {}} />);
    const file = screen.getByTestId('file');
    if (file) fireEvent.change(file, { target: { files: [new File(['test file'], 'test.geojson')] } });
    await waitFor(() => {
      expect(screen.getByText(errorMessage)).toBeInTheDocument();
    });
  });
});
