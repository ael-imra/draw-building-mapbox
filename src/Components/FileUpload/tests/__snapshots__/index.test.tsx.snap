// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1qf7f8o-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="file"
        id="file"
        type="file"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiGrid-root css-1onjoz5-MuiGrid-root"
  >
    title
  </div>
</DocumentFragment>
`;
